homeassistant/components/hlk_sw16.py

```diff
@@ -0,0 +1,294 @@
+            return
+
+        comps = {'switch': []}
+        for switch in switches.items():
+            comps['switch'].append(switch)
+
+        # Load platforms
+        for comp_name, comp_conf in comps.items():
+            if comp_conf:
+                load_platform(hass, comp_name, DOMAIN, {DOMAIN: comp_conf},
+                              config)
+
+        # Bind protocol to command class to allow entities to send commands
+        SW16Command.set_hlk_sw16_protocol(protocol)
+
+        # There is a valid connection to a HLK-SW16 device now so
+        # mark entities as available
+        async_dispatcher_send(hass, SIGNAL_AVAILABILITY, True)
+
+        # handle shutdown of HLK-SW16 asyncio transport
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP,
+                                   lambda x: transport.close())
+
+        _LOGGER.info('Connected to HLK-SW16')
+
+    hass.async_create_task(connect())
+    return True
+
+
+class SW16Device(Entity):
+    """Representation of a HLK-SW16 device.
+
+    Contains the common logic for HLK-SW16 entities.
+    """
+
+    platform = None
+    _available = True
+
+    def __init__(self, device_id, device_port, initial_event=None,
+                 name=None):
+        """Initialize the device."""
+        # HLK-SW16 specific attributes for every component type
+        self._initial_event = initial_event
+        self._device_id = device_id
+        self._device_port = str(device_port)
+        self._is_on = None
+        if name:
+            self._name = name
+        else:
+            self._name = device_id
+
+    @callback
+    def handle_event_callback(self, event):
+        """Propagate changes through ha."""
+        self.async_schedule_update_ha_state()
+
+    def _handle_event(self, event):
+        """Platform specific event handler."""
+        raise NotImplementedError()
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def name(self):
+        """Return a name for the device."""
+        return self._name
+
+    @property
+    def is_on(self):
+        """Return true if device is on."""
+        return self._is_on
+
+    @property
+    def available(self):
+        """Return True if entity is available."""
+        return self._available
+
+    @callback
+    def _availability_callback(self, availability):
+        """Update availability state."""
+        self._available = availability
+        self.async_schedule_update_ha_state()
+
+    async def async_added_to_hass(self):
+        """Register update callback."""
+        async_dispatcher_connect(self.hass, SIGNAL_AVAILABILITY,
+                                 self._availability_callback)
+
+        # Process the initial event now that the entity is created
+        if self._initial_event:
+            self.handle_event_callback(self._initial_event)
+
+
+class SW16Command(SW16Device):
```

Comment:
We shouldn't expose the protocol to home assistant at all. This class should move to the interface library, which should expose turn on, turn off and status methods as needed.

Reply:
SW16Device isn't the protocol and is home-assistant specific, [SW16Protocol](https://github.com/jameshilliard/hlk-sw16/blob/040c8cf6b0ce1028c4fe30fb4e13a083d34e082f/hlk_sw16/protocol.py#L9) is the protocol.

Reply:
`SW16Command` class should move to the interface library. The user of the library should somehow get an instance of this class, where turn on, turn off, get_status and is_connected are available as methods/properties.

Reply:
Hmm, I can refactor most parts of this into the interface library but I need to bind the protocol with `set_hlk_sw16_protocol` somewhere still I think, should I do that in `SW16Device` instead?

Reply:
The protocol should be an implementation detail of the library and somehow linked to instances of `SW16Command`. Home assistant don't need to know about the protocol.

Reply:
So `SW16Command` is mostly an abstraction over the individual relays while `SW16Protocol` in the interface library is an abstraction over all 16 device relays but `SW16Protocol` still has methods for accessing the relays individually, as it is right now `SW16Command` is just using those `SW16Protocol` methods that access the relays individually. I'm not really sure how I would move `SW16Command` into the interface library.

Reply:
Ok, it's ok to keep protocol as the interface point to the library. But then we can drop the `SW16Command` abstraction, and move protocol attribute to `SW16Device`.

Use our dispatch helper to update the protocol attribute with the new protocol on connection. So don't have set protocol be a class method.

Reply:
I've eliminated `SW16Command` and moved to callback based state changes.

> Use our dispatch helper to update the protocol attribute with the new protocol on connection. So don't have set protocol be a classmethod.

Is there an example or details for how to do this?

Reply:
Similar as to what we do here:
https://github.com/home-assistant/home-assistant/pull/17855/files#diff-4091ae1d5f6d8786d50478f1fed3f34aR78

Connect a method in the entity that sets the protocol argument to the protocol attribute. Send the new protocol with the dispatcher send function, using the correct signal.